### PR TITLE
Simple storage service attached to the window object of background.js

### PR DESCRIPTION
**Pull request by zhoutwo**

### CODE REVIEW

File: component/client/chat/Message.tsx [hunk 1]
```diff
@@ -38,26 +38,26 @@ class Message extends React.Component<IMessageProps, any> {
...
+    for (let i = 0; i < str.length; i++) {
+        hash = str.charCodeAt(i) + ((hash << 5) - hash);
+    }
+    return hash;
   }
 
   // Convert an int to hexadecimal with a max length
   // of six characters.
   // Source: https://www.designedbyaturtle.co.uk/2014/convert-string-to-hexidecimal-colour-with-javascript-vanilla/
   private intToARGB(i: number): string {
```
Comment (lamdaV) [inline component/client/chat/Message.tsx]: Why do you need this?

File: component/main.tsx [hunk 2]
```diff
@@ -16,3 +16,9 @@ const config = {
 firebase.initializeApp(config);
 
 ReactDOM.render(<ClientCore />, document.getElementById("root"));
+
+const storageService = (chrome.extension.getBackgroundPage() as any).BackgroundStorageService;
```
Comment (lamdaV) [inline component/main.tsx]: This should be in the store.

File: component/client/chat/Message.tsx [hunk 3]
```diff
@@ -38,26 +38,26 @@ class Message extends React.Component<IMessageProps, any> {
   // Then we'll use the int and convert to hex.
   // Source: https://www.designedbyaturtle.co.uk/2014/convert-string-to-hexidecimal-colour-with-javascript-vanilla/
   private hashCode(str: string): number {
```
Comment (lamdaV) [inline component/client/chat/Message.tsx]: Why is there change?

File: client/src/bg/background.js [hunk 4]
```diff
@@ -4,10 +4,51 @@
 //     "sample_setting": "This is how you use Store.js to remember values"
 // });
 
+function generateUUID() {
+  let id = 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function(c) {
+    let r = Math.random()*16|0, v = c === 'x' ? r : (r&0x3|0x8);
+    return v.toString(16);
+  });
+  return id;
+}
```
Comment (fredzqm) [inline client/src/bg/background.js]: Why use this to generate an ID?
Can you ask chrome to get one?

Comment (fredzqm) [inline client/src/bg/background.js]: https://developer.chrome.com/apps/app_identity
We can use identity of chrome API. That can serve as a unique ID>

File: client/src/bg/background.js [hunk 5]
```diff
@@ -4,10 +4,51 @@
...
+    data[key] = value;
+    this.storage.set(data);
+  }
+
+  remove(key) {
+    this.storage.remove(key);
+  }
+}
+
+window.BackgroundStorageService = new StorageService();
```
Comment (fredzqm) [inline client/src/bg/background.js]: Do we have to attach it to window?
Can't we use typescript import syntax?

File: component/main.tsx (hunk 2 above)
Comment (fredzqm) [inline component/main.tsx]: Is that the only way to get an object from the background?

File: client/src/bg/background.js (hunk 4 above)
Comment (zhoutwo) [inline client/src/bg/background.js]: We could, but I don't think it's necessary for the current state of the app. We can be agiler if we use a simple and relatively reliable method for now, and then switch to Chrome OAuth later.

There are people thinking the same thing: https://stackoverflow.com/questions/23822170/getting-unique-clientid-from-chrome-extension#answer-23854032

File: client/src/bg/background.js [hunk 6]
```diff
@@ -4,10 +4,51 @@
...
+
+  reset() {
+    this.storage.clear(() => {
+      defaults.id = generateUUID();
+      this.set(defaults);
+    });
+  }
+
+  get(key) {
+    return new Promise((resolve) => {
```
Comment (zhoutwo) [inline client/src/bg/background.js]: The API doesn't take an error callback, so there is nothing for me to reject with

Comment (zhoutwo) [inline client/src/bg/background.js]: The API doesn't take an error callback, so I have nothing to reject with.

File: client/src/bg/background.js (hunk 5 above)
Comment (zhoutwo) [inline client/src/bg/background.js]: TypeScript doesn't touch this part of the code given our current project structure.

File: component/client/chat/Message.tsx (hunk 3 above)
Comment (zhoutwo) [inline component/client/chat/Message.tsx]: There was double indent

File: component/main.tsx (hunk 2 above)
Comment (zhoutwo) [inline component/main.tsx]: Store as in the Redux Store?
This is how you retrieve the service from background.js

Comment (zhoutwo) [inline component/main.tsx]: The easiest way.

File: client/src/bg/background.js (hunk 4 above)
Comment (zhoutwo) [inline client/src/bg/background.js]: I think right now this serves our purpose and allows us to move on. Using Chrome's user identify involving touch more areas, and this UUID is good unless we have millions of users. We can make the switch later and be agiler right now.

There are others who think the same: https://stackoverflow.com/questions/23822170/getting-unique-clientid-from-chrome-extension#answer-23854032

File: component/main.tsx (hunk 2 above)
Comment (lamdaV) [inline component/main.tsx]: I would prefer that any required information that we need from the Chrome's local storage be done via Redux action. This allows us to organize our logic into a frontend component only caring about displaying data and a backend component only caring about retrieval and processing of data.

**Comment by zhoutwo** [comment]
Got approval offline to merge

**Pull request merged**